apps/ngrx-1/src/app/app.selectors.ts

```diff
@@ -0,0 +1,34 @@
+import { createSelector } from '@ngrx/store';
+import { selectUserIsAdmin } from './store/user/user.selectors';
+import { selectActivities } from './store/activity/activity.selectors';
+import { ActivityType, Person } from './store/activity/activity.model';
+
+const selectTeachersByActivityMap = createSelector(
+  selectUserIsAdmin,
+  selectActivities,
+  (isAdmin, activities) => {
+    if (isAdmin) {
+      return activities.reduce((statusMap, activity) => {
+        const status = statusMap[activity.type];
+        if (status) {
+          status.push(activity.teacher);
+        } else {
+          statusMap[activity.type] = [activity.teacher];
+        }
+        return statusMap;
+      }, {} as Record<ActivityType, Person[]>);
+    }
+    return {} as Record<ActivityType, Person[]>;
+  }
+);
```

Comment:
to avoir some nesting and make it more complicated to read, you should treat the case of notAdmin first
like 

```ts
if(!isAdmin) return blablabl

// case isAdmin 
```

try also to avoid 'as', it's better to type the return of the function

```ts
(isAdmin, actitivities):  Record<ActivityType, Person[]> => {
 return activities.reduce((statusMap:Record<ActivityType, Person[]>, activity) => {
 // do stuff 
}

}
```

Reply:
Nice idea with the early termination if statement.

I don't think it's possible to avoid the `as` type overrides? Because I need to type the `{}` initial value of the reducer. 

So I can simplify it to this:
```
(isAdmin, activities) => {
    if (!isAdmin) return {};

    return activities.reduce((statusMap, activity): Record<ActivityType, Person[]> => {
      const status = statusMap[activity.type];
      if (status) {
        status.push(activity.teacher);
      } else {
        statusMap[activity.type] = [activity.teacher];
      }
      return statusMap;
    }, {} as Record<ActivityType, Person[]>);
  }
```

Any ideas?

Reply:
maybe you are right. This is fine for me. 😊